### ISSUE TITLE
`util/http`: Cleanup

### ISSUE THREAD

**Issue by viceice** [renovatebot/renovate#11450]
### Describe the proposed change(s).

- [x] #14335 Move interfaces and types from `index.ts` to  `types.ts`
- [ ] #41293
- [ ] Split renovate options from `got` options, so they don't bubble to `got` where they can maybe harm
- [ ] #9651

**Comment by MaronHatoum** [comment]
Hi
can you please assign this issue to me?

**Comment by viceice** [comment]
This is a little bit more complicated, so should be done in multiple Pr's. Start with types move.

Then we can do the options refactoring, which will hit a lot datasources.

**Comment by MaronHatoum** [comment]
Hi @viceice,
I already did the first phase(types move) and I will create PR soon.
what should I do in the second phase?

**Comment by viceice** [comment]
> Split renovate options from got options, so they don't bubble to got where they can maybe harm
> 
> about the second phase solution: 
> - I should move all the `got` definitions to the `GotContextOptions` object?
> 
> **That means to move this part:**
> https://github.com/renovatebot/renovate/blob/15c393f79115c0a5fe6429396c5a9e51127ffc67/lib/util/http/types.ts#L12-L14
> 
> **To:**
> https://github.com/renovatebot/renovate/blob/15c393f79115c0a5fe6429396c5a9e51127ffc67/lib/util/http/types.ts#L7-L9
> 
> After moving `gotOptions` I will fix all the usages.
> Is it the right way to make this refactor?
> 

No, we should move got options to a child object of renovate http options.

**Comment by MaronHatoum** [comment]
I have 2 different ideas:
1. to implement a new interface I will call it `GotOptions` which extends `HttpOptions`:
https://github.com/renovatebot/renovate/blob/36a29974bbf94251f640e3c06be816b16a79c252/lib/util/http/index.ts#L26-L40
and the new interface will look like this:
`export interface GotOptions extends HttpOptions {`
`GotOptions: GotBufferOptions | GotJSONOptions;`
`}`

2. just to add the `GotOptions` object to the HttpOptions interface like:

`export interface HttpOptions {`
  `body?: any;`
  `username?: string;`
  `password?: string;`
  `baseUrl?: string;`
  `headers?: OutgoingHttpHeaders;`
  `noAuth?: boolean;`
 ` throwHttpErrors?: boolean;`
  `useCache?: boolean;`

  `GotOptions: GotBufferOptions | GotJSONOptions;`
`}`

please let me know which change do you prefer.

**Comment by viceice** [comment]
Second Variant

**Comment by MaronHatoum** [comment]
Hi @viceice 
regarding the third part of this refactor (remove legacy.ts)
should I remove `legacy.ts` with the properties and their usages?
or 
to move the properties to another place then remove to `legacy.ts` and change the property usages to the new properties place? 
thanks.

**Comment by viceice** [comment]
> Hi @viceice 
> regarding the third part of this refactor (remove legacy.ts)
> should I remove `legacy.ts` with the properties and their usages?
> or 
> to move the properties to another place then remove to `legacy.ts` and change the property usages to the new properties place? 
> thanks.

Remove, but you need to check the whole codebase for usage. Typescript don't know all of them. Also check the `global.d.ts` file for obsolete `Error` properties.

**Comment by MaronHatoum** [comment]
Hi @viceice 
regarding the third part of this task remove `legacy.ts`
should I remove this field from `GotExtraOptions` too ?

https://github.com/renovatebot/renovate/blob/1684b1c3e14e74aa1d47fe50e42bf8a020ae446a/lib/util/http/types.ts#L18
